### DISCUSSION TITLE
Does getStaticProps have access to context or not?

### DISCUSSION THREAD

**Discussion by chhuang** [vercel/next.js#49568] [Help] [answered]
### Summary

So on [this page](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props#when-does-getstaticprops-run) it says:

> getStaticProps does not have access to the incoming request (such as query parameters or HTTP headers) as it generates static HTML.

But then on the [API reference page](https://nextjs.org/docs/pages/api-reference/functions/get-static-props):

```
export async function getStaticProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  };
}
```

> The context parameter is an object containing the following keys:
>
> `params` contains the route parameters for pages using [dynamic routes](https://nextjs.org/docs/pages/building-your-> application/routing/dynamic-routes). For example, if the page name is [id].js , then params will look like { id: ... }. You should > use this together with getStaticPaths, which we’ll explain later.

For the later, I assume I have access to `context.params.postId` inside `getStaticProps` for `/pages/posts/[postId].tsx`. So why does the docs say it does not have access? Is the context object not part of the incoming request?

### Additional information

I've tested this and I do have access to `context.params.postId` in `getStaticProps`, I'm just a bit confused with what the docs say about not having access to the incoming request.

**Comment by icyJoseph** [comment]
Hi,

`getStaticProps`, unlike `getServerSideProps`, never interacts with incoming requests.

Let's talk about production builds first.

When you run `next build`, if a page uses `getStaticProps`, Next.js executes the function and creates props for the page, at build time. There's no server running yet! So it has no knowledge of any incoming request, because there's none. 

The build process outputs HTML/CSS/JS to drive this page, and that is it. The page has been pre-rendered.

When the page is requests, Next.js serves the static assets generated during 'pre-rendering'.

Now if you have activated ISR (Incremental Static Regeneration), at runtime, when you run `next start`, given the ISR rules, Next.js re-runs `getStaticProps`, but it doesn't do it on request, it does it following the ISR revalidation time.

In the other hand `getServerSideProps`, doesn't do anything during `next build`, no page is built. Later during `next start`, as a request comes in, it generates the page's HTML/CSS/JS. During this process `getServerSideProps` has access to the incoming requests and outgoing response.

In development mode, `gSP` and `gSSP` run on every request and Hot Module Reload. Imagine if you had to restart the dev server, to re-run `gSP`, that'd be terrible.

**Reply by chhuang** [reply]
So the reason I have access to `context.params.postId` in `getStaticProps` is because I'm running it locally?

**Reply by icyJoseph** [reply, marked as answer]
No, `context.params.postId` is passed to the the `getStaticProps` function during build time and dev mode. How did you arrive to that conclusion? I'd like to make my first comment more precise.

Like, I think you are assuming context, to be a req/res pair kind of thing, but that's not the case for `getStaticProps`.

The context signature for `gSP` in all modes, is:

```ts
/**
 * Context object passed into `getStaticProps`.
 * @link https://nextjs.org/docs/api-reference/data-fetching/get-static-props#context-parameter
 */
export type GetStaticPropsContext<
  Params extends ParsedUrlQuery = ParsedUrlQuery,
  Preview extends PreviewData = PreviewData
> = {
  params?: Params
  preview?: boolean
  previewData?: Preview
  draftMode?: boolean
  locale?: string
  locales?: string[]
  defaultLocale?: string
}
```

There's no request, no response, the framework only gives you access to those. Most notable, params, which contains the dynamic parts of the URL. Which in turn is paired with `getStaticPaths`, and the `ISR` feature.

Again, when `GSP` runs, there's no incoming request, not outgoing response. I think you might be placing some expectation in the `context` term. It doesn't imply incoming requests, nor outgoing response.

**Reply by chhuang** [reply]
Right! So the context in `getStaticProps = async (context) => {}` is not exactly the same object as the one in `getServersideProps = async (context) => {}`. Even though they both have access to `context.params`. 

> getStaticProps does not have access to the incoming request (such as query parameters or HTTP headers) as it generates static HTML.

In this case, does it mean that `postId` is not considered as a query param? I think I may be confusing between `context.params.postId` with `context.query.postId` (because of the wordings "query parameters").

<img width="512" alt="Screenshot 2023-05-10 at 5 49 53 pm" src="https://github.com/vercel/next.js/assets/78710/b7a3f717-b81d-49fb-a6eb-3d90b880a1ed">

<img width="510" alt="Screenshot 2023-05-10 at 5 49 37 pm" src="https://github.com/vercel/next.js/assets/78710/46efb7bd-fbc9-41ba-8661-31469b24b8e9">
